[SPARK-21259] More rules for scalastyle

gengliangwang opened the pull request:
## What changes were proposed in this pull request?
Add more rules for scalastyle

During code review, we spent so much time on code style issues.
It would be great if we add rules:
1) disallow space before colon
2) disallow space before right parentheses
3) disallow space after left parentheses

## How was this patch tested?

unit tests, integration tests

Please review http://spark.apache.org/contributing.html before opening a pull request.

gengliangwang committed:
first run

gengliangwang committed:
resolve hard ones

gengliangwang committed:
fix one error

gengliangwang committed:
resolve more

gengliangwang committed:
more scalastyle off/on

gengliangwang committed:
Insert a space after the start of the comment

SparkQA commented:
**[Test build #78937 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78937/testReport)** for PR 18471 at commit [`f68e600`](https://github.com/apache/spark/commit/f68e60014ca2f1dbb502dbb942e109505c892c5a).

SparkQA commented:
**[Test build #78940 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78940/testReport)** for PR 18471 at commit [`d1b4bbe`](https://github.com/apache/spark/commit/d1b4bbe46a279959816e57a9c16636dd5e2d23ad).

gengliangwang commented:
Most changes are done by my ruby script:
```ruby
lines = `./dev/lint-scala 2>&1 | grep "before\\|after"`.split("\n")
lines.each do |line|
  file, line_num=line.split(/[ ,:]/)[1,2]
  if line =~ /Space before token :/
    `sed -i '' '#{line_num}s/ :/:/g' #{file}`
  elsif line =~ /Space before token \)/
    `sed -i '' '#{line_num}s/ )/)/g' #{file}`
  elsif line =~ /Space after token \(/
    `sed -i '' '#{line_num}s/( /(/g' #{file}`
  end
end
```

cloud-fan commented on core/src/main/scala/org/apache/spark/Accumulable.scala:
```diff
@@ -68,7 +68,9 @@ class Accumulable[R, T] private (
 
   private[spark] def this(initialValue: R, param: AccumulableParam[R, T], name: Option[String]) = {
+    // scalastyle:off
     this(initialValue, param, name, false /* countFailedValues */)
```
nit: `countFailedValues  = false` is better, this style is only useful for java

HyukjinKwon commented on scalastyle-config.xml:
```diff
@@ -46,7 +46,7 @@ This file is divided into 3 sections:
   <check level="error" class="org.scalastyle.file.FileTabChecker" enabled="true"></check>
 
-  <check level="error" class="org.scalastyle.file.HeaderMatchesChecker" enabled="true">
+  <check level="error" class="org.scalastyle.file.HeaderMatchesChecker" enabled="false">
```
Just a question. Should we disable this rule to enable the new rule?

gengliangwang replied:
No, I made a mistake.

HyukjinKwon commented on scalastyle-config.xml:
```diff
@@ -245,7 +245,13 @@ This file is divided into 3 sections:
+    </parameters>
+  </check>
+
+  <check level="error" class="org.scalastyle.scalariform.DisallowSpaceAfterTokenChecker" enabled="true">
```
Another question. Could we add `customId` and disable this only when `scalastyle:off` and `scalastyle:on` is required?

gengliangwang replied:
Thanks, I have updated the xml file.

SparkQA commented:
**[Test build #78937 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78937/testReport)** for PR 18471 at commit [`f68e600`](https://github.com/apache/spark/commit/f68e60014ca2f1dbb502dbb942e109505c892c5a).
 * This patch **fails Spark unit tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78937/
Test FAILed.

SparkQA commented:
**[Test build #78940 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78940/testReport)** for PR 18471 at commit [`d1b4bbe`](https://github.com/apache/spark/commit/d1b4bbe46a279959816e57a9c16636dd5e2d23ad).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78940/
Test PASSed.

gengliangwang committed:
fix mistake: enable org.scalastyle.file.HeaderMatchesChecker

SparkQA commented:
**[Test build #78948 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78948/testReport)** for PR 18471 at commit [`05084a9`](https://github.com/apache/spark/commit/05084a9706b2c3a9cf6dc1c733d7d16b8bac0f41).

gengliangwang committed:
add customId for new rule; scala files will be updated later

SparkQA commented:
**[Test build #78951 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78951/testReport)** for PR 18471 at commit [`fd5bde5`](https://github.com/apache/spark/commit/fd5bde5e1c79c68bebd667828d773f63165117cb).

SparkQA commented:
**[Test build #78948 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78948/testReport)** for PR 18471 at commit [`05084a9`](https://github.com/apache/spark/commit/05084a9706b2c3a9cf6dc1c733d7d16b8bac0f41).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
**[Test build #78951 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78951/testReport)** for PR 18471 at commit [`fd5bde5`](https://github.com/apache/spark/commit/fd5bde5e1c79c68bebd667828d773f63165117cb).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78951/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78948/
Test FAILed.

srowen commented:
I tend to agree with the style change, but, am concerned about what a huge diff this creates for little gain. I personally am not sure I'd bother with the colon change; some instances of space before colon are conventional like `[K : ClassTag]`, although many you fixed should indeed not have had a space. I also am not sure we want to enforce the naming convention you suggest here.

The space inside parens change seems OK, maybe smaller.

HyukjinKwon commented on core/src/main/scala/org/apache/spark/broadcast/BroadcastManager.scala:
```diff
@@ -52,8 +52,8 @@ private[spark] class BroadcastManager(
 
-  def newBroadcast[T: ClassTag](value_ : T, isLocal: Boolean): Broadcast[T] = {
-    broadcastFactory.newBroadcast[T](value_, isLocal, nextBroadcastId.getAndIncrement())
+  def newBroadcast[T: ClassTag](_value: T, isLocal: Boolean): Broadcast[T] = {
```
I am not sure of this naming change. I mean the change does not look particulaily better to me. I believe we have a open PR - https://github.com/databricks/scala-style-guide/pull/52 which appearntly stays against some cases here.

cloud-fan replied:
I think it's not about adding `_` to the name of private val, but kind of avoiding name conflict if 2 variables have similar meaning. At least this is fine to me, I also do this a lot...

HyukjinKwon replied:
Yea, not this one but some cases look provate ones. I am fine if we prefer a specific way. My only worry is we happen to fix it back again in the future but no one stays against because they don't know which way is preferred...

HyukjinKwon replied:
Maybe, this can be an example other contributors like me can refer in the future BTW.

HyukjinKwon commented on core/src/main/scala/org/apache/spark/Accumulable.scala:
```diff
@@ -68,7 +68,9 @@ class Accumulable[R, T] private (
   private[spark] def this(initialValue: R, param: AccumulableParam[R, T], name: Option[String]) = {
+    // scalastyle:off
     this(initialValue, param, name, false /* countFailedValues */)
+    // scalastyle:on
```
I believe we could enable and disable via `scalastyle:on customid`  rather than switching off whole style rules here IIRC.

HyukjinKwon replied:
Oh, nvm. I just read your commit log ...

cloud-fan commented:
My only concern is `[K : ClassTag]`. This is conventional in scala but the new style check rule rejects it. If it's hard to make a corrected rule for this case, we can drop it.

jiangxb1987 commented:
Do we still want this?

gengliangwang commented:
Close this for now.
